test/unit/cmap/auth/mongodb_oidc/callback_lock_cache.test.ts

```diff
@@ -0,0 +1,144 @@
+import { expect } from 'chai';
+import * as sinon from 'sinon';
+import { setTimeout } from 'timers';
+
+import {
+  CallbackLockCache,
+  Connection,
+  MongoCredentials,
+  MongoInvalidArgumentError
+} from '../../../../mongodb';
+
+describe('CallbackLockCache', function () {
+  describe('#getCallbacks', function () {
+    const connection = sinon.createStubInstance(Connection);
+    connection.address = 'localhost:27017';
+
+    context('when a request callback does not exist', function () {
+      const credentials = new MongoCredentials({
+        username: 'test_user',
+        password: 'pwd',
+        source: '$external',
+        mechanismProperties: {}
+      });
+      const cache = new CallbackLockCache();
+
+      it('raises an error', function () {
+        try {
+          cache.getCallbacks(connection, credentials);
+        } catch (error) {
+          expect(error).to.be.instanceOf(MongoInvalidArgumentError);
+          expect(error.message).to.include(
+            'Auth mechanism property REQUEST_TOKEN_CALLBACK is required'
+          );
+        }
+      });
+    });
+
+    context('when no entry exists in the cache', function () {
+      context('when a refresh callback exists', function () {
+        let requestCount = 0;
+        let refreshCount = 0;
+
+        const request = async () => {
+          requestCount++;
+          if (requestCount > 1) {
+            throw new Error('Cannot execute request simultaneously.');
+          }
+          await new Promise(resolve => setTimeout(() => resolve(), 1000));
+          requestCount--;
+          return Promise.resolve({ accessToken: '' });
+        };
+        const refresh = async () => {
+          refreshCount++;
+          if (refreshCount > 1) {
+            throw new Error('Cannot execute refresh simultaneously.');
+          }
+          await new Promise(resolve => setTimeout(() => resolve(), 1000));
+          refreshCount--;
+          return Promise.resolve({ accessToken: '' });
+        };
+        const requestSpy = sinon.spy(request);
+        const refreshSpy = sinon.spy(refresh);
+        const credentials = new MongoCredentials({
+          username: 'test_user',
+          password: 'pwd',
+          source: '$external',
+          mechanismProperties: {
+            REQUEST_TOKEN_CALLBACK: requestSpy,
+            REFRESH_TOKEN_CALLBACK: refreshSpy
+          }
+        });
+        const cache = new CallbackLockCache();
+        const { requestCallback, refreshCallback, callbackHash } = cache.getCallbacks(
+          connection,
+          credentials
+        );
+
+        it('puts a new entry in the cache', function () {
+          expect(cache.entries.size).to.equal(1);
```

Comment:
nit: you can use `lengthOf` assertions on maps / sets - it will print correctly:
`expected Map{} to have a size of 1 but got +0`

Reply:
Changed all these to use `lengthOf` assertions.